File: TCO 2016 Round 2A/LCMGCD.java
import java.io.*;
import java.util.*;

public class LCMGCD {
    private static final String YES = "Possible";
    private static final String NO = "Impossible";
    int n;
    int signum(int x) {
        if (x == 0) return 0;
        return x > 0 ? 1 : -1;
    }
    int cmp(int x, int y) {
        return 1 + signum(x - y);
    }
    boolean can(int[][] a) {
		if (a[1][1] > 0) {
		  if (a[0][2] != 0 && n == a[0][2] + 1) return false;
		  if (a[2][0] != 0 && n == a[2][0] + 1) return false;
		  return true;
		}
		if (a[0][1] > 0 && a[1][0] > 0) {
		  return true;
		}
		if (a[2][1] > 0 && a[1][2] > 0) {
		  return true;
		}
		if (a[2][1] > 0 && a[1][0] > 0) {
		  if (a[0][2] > 0) return true;
		  return false;
		}
		if (a[0][1] > 0 && a[1][2] > 0) {
		  if (a[2][0] > 0) return true;
		  return false;
		}
		return false;
    }
	public String isPossible(int[] x, int t) {
        int[][] a = new int[3][3];
        this.n = x.length;
        Pair o = new Pair(t);
        for (int val : x) {
            Pair pair = new Pair(val);
            ++a[cmp(o.x, pair.x)][cmp(o.y, pair.y)];
        }
		return can(a) ? YES : NO;
	}
    static class Pair {
        int x, y;
        public Pair(){}
        public Pair(int val) {
            x = y = 0;
            while (val % 2 == 0 && val != 1) {
                val /= 2;
                ++x;
            }
            while (val % 3 == 0 && val != 1) {
                val /= 3;
                ++y;
            }
        }
    }

// CUT begin
	public static void main(String[] args){
		System.err.println("LCMGCD (400 Points)");
		System.err.println();
		HashSet<Integer> cases = new HashSet<Integer>();
        for (int i = 0; i < args.length; ++i) cases.add(Integer.parseInt(args[i]));
        runTest(cases);
	}

	static void runTest(HashSet<Integer> caseSet) {
	    int cases = 0, passed = 0;
	    while (true) {
	    	String label = Reader.nextLine();
	    	if (label == null || !label.startsWith("--"))
	    		break;

            int[] x = new int[Integer.parseInt(Reader.nextLine())];
            for (int i = 0; i < x.length; ++i)
                x[i] = Integer.parseInt(Reader.nextLine());
            int t = Integer.parseInt(Reader.nextLine());
            Reader.nextLine();
            String __answer = Reader.nextLine();

            cases++;
            if (caseSet.size() > 0 && !caseSet.contains(cases - 1))
                continue;
    		System.err.print(String.format("  Testcase #%d ... ", cases - 1));

            if (doTest(x, t, __answer))
                passed++;
	    }
	    if (caseSet.size() > 0) cases = caseSet.size();
        System.err.println(String.format("%nPassed : %d/%d cases", passed, cases));

        int T = (int)(System.currentTimeMillis() / 1000) - 1499514490;
        double PT = T / 60.0, TT = 75.0;
        System.err.println(String.format("Time   : %d minutes %d secs%nScore  : %.2f points", T / 60, T % 60, 400 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT))));
	}

	static boolean doTest(int[] x, int t, String __expected) {
		long startTime = System.currentTimeMillis();
		Throwable exception = null;
		LCMGCD instance = new LCMGCD();
		String __result = "";
		try {
			__result = instance.isPossible(x, t);
		}
		catch (Throwable e) { exception = e; }
		double elapsed = (System.currentTimeMillis() - startTime) / 1000.0;

		if (exception != null) {
			System.err.println("RUNTIME ERROR!");
			exception.printStackTrace();
			return false;
		}
		else if (__expected.equals(__result)) {
			System.err.println("PASSED! " + String.format("(%.2f seconds)", elapsed));
			return true;
		}
		else {
			System.err.println("FAILED! " + String.format("(%.2f seconds)", elapsed));
			System.err.println("           Expected: " + "\"" + __expected + "\"");
			System.err.println("           Received: " + "\"" + __result + "\"");
			return false;
		}
	}

	static class Reader {
        private static final String dataFileName = "LCMGCD.sample";
	    private static BufferedReader reader;

	    public static String nextLine() {
	        try {
                if (reader == null) {
                    reader = new BufferedReader(new InputStreamReader(new FileInputStream(dataFileName)));
                }
                return reader.readLine();
	        }
	        catch (IOException e) {
	            System.err.println("FATAL!! IOException");
	            e.printStackTrace();
	            System.exit(1);
	        }
	        return "";
	    }
	}
// CUT end
}

File: TCO 2016 Round 2A/LCMGCD.cpp
#include <cstdio>
#include <cmath>
#include <cstring>
#include <ctime>
#include <iostream>
#include <algorithm>
#include <set>
#include <vector>
#include <sstream>
#include <typeinfo>
#include <fstream>

using namespace std;
typedef pair<int,int> PII;
#define fi first
#define se second
PII build_gcd(PII a, PII b) {
    return PII(max(a.fi, b.fi), max(a.se, b.se));
}
PII build_lcm(PII a, PII b) {
    return PII(min(a.fi, b.fi), min(a.se, b.se));
}
PII calc(int n) {
    int k1 = 0, k2 = 0;
    while (n % 2 == 0 && n != 1) n /= 2, ++k1;
    while (n % 3 == 0 && n != 1) n /= 3, ++k2;
    return PII(k1, k2);
}
class LCMGCD {
    public:
    string isPossible(vector<int> x, int t) {
        time_t start = clock();
        PII target = calc(t);
        vector<PII> a;
        for (int tmp : x) {
            a.push_back(calc(tmp));
        }
        for (int it = 0; ; ++it) {
        //    cout << "cnt=" << it << endl;
            vector<PII> v;
            v.push_back(a[0]);
            for (int i = 1; i < a.size(); ++i) {
                vector<PII> nv;
                for (int j = 0; j < v.size(); ++j) {
                    nv.push_back(build_lcm(v[j], a[i]));
                    nv.push_back(build_gcd(v[j], a[i]));
                }
                sort(nv.begin(), nv.end());
                nv.resize(distance(nv.begin(), unique(nv.begin(), nv.end())));
                v=nv;
            }
            //if (binary_search(v.begin(), v.end(), target)) return "Possible";
            vector<PII>::iterator iter = lower_bound(v.begin(), v.end(), target);
            if (iter != v.end() && (*iter == target)) return "Possible";
            int tmp = 63 & it;
            if (tmp == 0 && clock() - start > 0.9 * CLOCKS_PER_SEC) {
               break;
            }
            random_shuffle(a.begin(), a.end());
        }
        return "Impossible";
    }
};

// CUT begin
ifstream data("LCMGCD.sample");

string next_line() {
    string s;
    getline(data, s);
    return s;
}

template <typename T> void from_stream(T &t) {
    stringstream ss(next_line());
    ss >> t;
}

void from_stream(string &s) {
    s = next_line();
}

template <typename T> void from_stream(vector<T> &ts) {
    int len;
    from_stream(len);
    ts.clear();
    for (int i = 0; i < len; ++i) {
        T t;
        from_stream(t);
        ts.push_back(t);
    }
}

template <typename T>
string to_string(T t) {
    stringstream s;
    s << t;
    return s.str();
}

string to_string(string t) {
    return "\"" + t + "\"";
}

bool do_test(vector<int> x, int t, string __expected) {
    time_t startClock = clock();
    LCMGCD *instance = new LCMGCD();
    string __result = instance->isPossible(x, t);
    double elapsed = (double)(clock() - startClock) / CLOCKS_PER_SEC;
    delete instance;

    if (__result == __expected) {
        cout << "PASSED!" << " (" << elapsed << " seconds)" << endl;
        return true;
    }
    else {
        cout << "FAILED!" << " (" << elapsed << " seconds)" << endl;
        cout << "           Expected: " << to_string(__expected) << endl;
        cout << "           Received: " << to_string(__result) << endl;
        return false;
    }
}

int run_test(bool mainProcess, const set<int> &case_set, const string command) {
    int cases = 0, passed = 0;
    while (true) {
        if (next_line().find("--") != 0)
            break;
        vector<int> x;
        from_stream(x);
        int t;
        from_stream(t);
        next_line();
        string __answer;
        from_stream(__answer);

        cases++;
        if (case_set.size() > 0 && case_set.find(cases - 1) == case_set.end())
            continue;

        cout << "  Testcase #" << cases - 1 << " ... ";
        if ( do_test(x, t, __answer)) {
            passed++;
        }
    }
    if (mainProcess) {
        cout << endl << "Passed : " << passed << "/" << cases << " cases" << endl;
        int T = time(NULL) - 1499492393;
        double PT = T / 60.0, TT = 75.0;
        cout << "Time   : " << T / 60 << " minutes " << T % 60 << " secs" << endl;
        cout << "Score  : " << 400 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT)) << " points" << endl;
    }
    return 0;
}

int main(int argc, char *argv[]) {
    cout.setf(ios::fixed, ios::floatfield);
    cout.precision(2);
    set<int> cases;
    bool mainProcess = true;
    for (int i = 1; i < argc; ++i) {
        if ( string(argv[i]) == "-") {
            mainProcess = false;
        } else {
            cases.insert(atoi(argv[i]));
        }
    }
    if (mainProcess) {
        cout << "LCMGCD (400 Points)" << endl << endl;
    }
    return run_test(mainProcess, cases, argv[0]);
}
// CUT end
